File: 4th year/test2.cpp
#include<iostream>
#include<cmath>
#include<fstream>
using namespace std;

double f1(double t , double N)
{
       return  - 0.08*N;
  //  return N*exp(-0.08*t);
}

void p(double t0, double N0)
{
    ofstream fout("n.dat");
    ofstream file("n2.dat");
    double N=N0, k1,k2,k3,k4, h=0.0001;

    for(double t=t0; t<100; t+=0.001)
    {
        fout << t << "  " << log(N)  << "  "  <<  log(N0-N) << endl;
        file << t << "  " << N << "  " << N0-N <<  endl;
        
        k1 = h*f1(t,N);
        k2 = h*f1(t+h/2, N+k1/2);
        k2 = h*f1(t+h/2, N+k2/2);
        k4 = h*f1(t+h, N+k3);
        
        N += (k1 + 2*k2 + 2*k3 + k4)/6;
                

        //N += h*f1(t,N);
    }


}

int main()
{
    p(0,100000);
    return 0;
}

File: masters/test2.py
import numpy as np
import matplotlib.pyplot as plt
import math
from joblib import Parallel, delayed
import cmath

s1 = 0
s2 = 0
a = []
b = []
r = 10
T = 40

def x(n, m):
    if n == m:
        return 1/2
    else:
        return ((1 - pow(-1, n + m)) / (pow(math.pi, 2))) * ((1 / pow(n + m, 2)) - (1 / pow(n - m, 2)))

def E(n, m):
    return pow(math.pi, 2) * (pow(n, 2) - pow(m, 2))

def compute_s2(t):
    C = 0
    Z = 0
    for n in range(0, r):
        c = 0
        for m in range(0, r):
            b = 0
            for k in range(0, r):
                b += (1/2.0) * x(n, k) * x(k, m) * (E(k, m) * cmath.exp(1j * E(n, k) * t) - E(n, k) * cmath.exp(1j * E(k, m) * t))
            c += b * np.conjugate(b)

        Z += cmath.exp(-E(n,0)/T)
        C += (cmath.exp(-E(n,0)/T)*c)
    return C/Z
    


# Parallel execution using joblib
results = Parallel(n_jobs=-1)(delayed(compute_s2)(t) for t in np.arange(0, 1, 0.001))

a = np.arange(0, 1, 0.001)
b = results

data = np.column_stack((a, b))
np.savetxt('test.dat', data, delimiter='  ')

plt.plot(a, b)
plt.show()
